### PR TITLE
fix(core): adding accessability label and reading button type

### PR DESCRIPTION
**Pull request by Lokanathan-k**
#### Please provide a link to the associated issue.
Closes: https://github.com/SAP/fundamental-ngx/issues/6206

#### Please provide a brief summary of this pull request.

adding accessability label and reading button type

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [NA] update `README.md`
- [NA] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Review by Betrozov — APPROVED** [review]

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
In my opinion the aria-label should be an input on the button (`@Input() ariaLabel: string = null`).  And if the developer has not provided a value for this input, then the button could automatically set the `aria-label` using the value given by the developer for `label` (and the `fdType`, if provided. They should be comma-separated).  If the dev has provided the input `ariaLabel` then use that

**Comment by Lokanathan-k** [comment]
> provided a value for this input, th

@InnaAtanasova @mikerodonnell89 
There are few button with no label meaning the button with only icon in that case what should be done? if that is not handled aria label is something like this aria-label="undefined, emphasised"

**Comment by mikerodonnell89** [comment]
> There are few button with no label meaning the button with only icon in that case what should be done? if that is not handled aria label is something like this aria-label="undefined, emphasised"

Good question, I didn't think of this, maybe use the property passed in to the glyph input, instead of the label, thoughts @InnaAtanasova ? This should simply be the default behavior and the docs should make it clear to the developer that providing the appropriate label is their responsibility

**Comment by InnaAtanasova** [comment]
> > There are few button with no label meaning the button with only icon in that case what should be done? if that is not handled aria label is something like this aria-label="undefined, emphasised"
> 
> Good question, I didn't think of this, maybe use the property passed in to the glyph input, instead of the label, thoughts @InnaAtanasova ? This should simply be the default behavior and the docs should make it clear to the developer that providing the appropriate label is their responsibility

It was always the case. This is the current state of the buttons:
Screen Shot 2021-09-01 at 12 10 21 PM

The ones with no label (text) have `aria-label`.

**Comment by InnaAtanasova** [comment]
For this case, where we have label (text):
```
<button fd-button [fdMenu]="true" [fdType]="'positive'" [glyph]="'filter'" label="Filter"></button>
```

We can check if there's aria-label provided by the user and if not set it to the label text behind the scene.
If there's aria-label, do nothing.

In this case, where you have only icon, the user must provide aria-label, just like they did before:
```
<button fd-button [fdMenu]="true" [fdType]="'positive'" [glyph]="'accept'" aria-label="accept" title="accept"></button>
```

**Review by mikerodonnell89 — APPROVED** [review]

**Comment by Seamoo13** [comment]
pushed a fix for 1 test, the other failing test is due to an issue in your PR

1 issue causing core flexible column layout test to fail -  aria-label for button in dynamic example not present until after clicking the button

**STR:**

1. go to https://deploy-preview-6385--fundamental-ngx.netlify.app/#/core/flexible-column-layout
2. go to the 2nd example (Flexible Column Layout With Dynamic Page)
3. click to open the example
4. click 'open column 2' button
5. inspect the button with the icon of a pin
6. notice that there is no aria-label (on main branch there is)
7. click the pin button
8. notice that the aria-label has appeared
9. click button again
10. aria-label present

**Actual results:** pin button missing aria-label until after being clicked

**Expected results:** button should have aria-label on load

https://user-images.githubusercontent.com/47522152/132020633-d2374dfc-0fc5-4909-9130-c44e6375a6c2.mp4

File: libs/core/src/lib/bar/button-bar/button-bar.component.ts [hunk 1]
```diff
@@ -23,8 +23,8 @@ import { ContentDensityService } from '@fundamental-ngx/core/utils';
               [compact]="compact"
               [fdType]="fdType"
               [label]="label"
-              [attr.title]="title"
-              [attr.aria-label]="ariaLabel"
+              [attr.title]="ariaLabel"
```
Comment (InnaAtanasova) [inline libs/core/src/lib/bar/button-bar/button-bar.component.ts:26]: Shouldn't this first check if the title prop is set and then assign it to ariaLabel? 
We should not take away the option for the user to set the value of title to what they want

File: libs/core/src/lib/bar/button-bar/button-bar.component.ts [hunk 2]
```diff
@@ -23,8 +23,8 @@ import { ContentDensityService } from '@fundamental-ngx/core/utils';
...
+              [ariaLabel]="ariaLabel"
```
Comment (InnaAtanasova) [inline libs/core/src/lib/bar/button-bar/button-bar.component.ts:27]: I'm confused. Why did you change [attr.aria-label] to [ariaLabel]?

**Review by InnaAtanasova — DISMISSED** [review]

File: libs/core/src/lib/button/button.component.ts [hunk 3]
```diff
@@ -37,7 +37,8 @@ import { applyCssClass } from '@fundamental-ngx/core/utils';
     changeDetection: ChangeDetectionStrategy.OnPush,
     host: {
         '[attr.type]': 'type',
-        '[attr.disabled]': '_disabled || null'
+        '[attr.disabled]': '_disabled || null',
+        '[attr.aria-label]': 'ariaLabel ? ariaLabel : label !== undefined ? label +","+ fdType : glyph === undefined || glyph === null  ? title : glyph.split("-").join(" ") +","+ fdType '
```
Comment (droshev) [inline libs/core/src/lib/button/button.component.ts:41]: why do we need this?

File: libs/core/src/lib/button/base-button.ts [hunk 4]
```diff
@@ -61,6 +61,10 @@ export class BaseButton {
     @Input()
     fdMenu = false;
 
+    /** adding native aria-label to the component */
+    @Input()
+    ariaLabel: string = null;
```
Comment (droshev) [inline libs/core/src/lib/button/base-button.ts:66]: can't we use the native attribute and avoid the creation iof an input

File: apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html [hunk 5]
```diff
@@ -1,7 +1,7 @@
-<button fd-button label="Standard Button"></button>
-<button fd-button label="Emphasized Button" [fdType]="'emphasized'"></button>
-<button fd-button label="Ghost Button" [fdType]="'ghost'"></button>
-<button fd-button label="Positive Button" [fdType]="'positive'"></button>
-<button fd-button label="Negative Button" [fdType]="'negative'"></button>
-<button fd-button label="Attention Button" [fdType]="'attention'"></button>
-<button fd-button label="Transparent Button" [fdType]="'transparent'"></button>
+<button fd-button label="Standard Button" ariaLabel="Standard"></button>
+<button fd-button label="Emphasized Button" fdType="emphasized" ariaLabel="Emphasized"></button>
```
Comment (droshev) [inline apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html:2]: those updates of the doc where `[fdType]` is replaced with `fdType` should be done in separate Pr

File: apps/docs/src/app/core/component-docs/button/examples/button-menu-example.component.html [hunk 6]
```diff
@@ -1,14 +1,14 @@
-<button fd-button [fdMenu]="true" [glyph]="'cart'" label="Add to Cart"></button>
-<button fd-button [fdMenu]="true" [fdType]="'emphasized'" [glyph]="'cart'" label="Add to Cart"></button>
-<button fd-button [fdMenu]="true" [fdType]="'ghost'" [glyph]="'cart'" label="Add to Cart"></button>
-<button fd-button [fdMenu]="true" [fdType]="'positive'" [glyph]="'filter'" label="Filter"></button>
-<button fd-button [fdMenu]="true" [fdType]="'negative'" [glyph]="'filter'" label="Filter"></button>
-<button fd-button [fdMenu]="true" [fdType]="'attention'" [glyph]="'filter'" label="Filter"></button>
-<button fd-button [fdMenu]="true" [fdType]="'transparent'" [glyph]="'filter'" label="Filter"></button>
+<button fd-button [fdMenu]="true" glyph="cart" label="Add to Cart" ariaLabel="Add to cart"></button>
```
Comment (droshev) [inline apps/docs/src/app/core/component-docs/button/examples/button-menu-example.component.html:1]: those updates of the doc where `[glyph]` is replaced with `glyph` should be done in separate PR

File: apps/docs/src/app/core/component-docs/button/examples/button-icons-example.component.html [hunk 7]
```diff
@@ -1,15 +1,15 @@
-<button fd-button label="Add to Cart" [glyph]="'cart'"></button>
-<button fd-button label="Add to Cart" [fdType]="'emphasized'" [glyph]="'cart'"></button>
-<button fd-button label="Add to Cart" [fdType]="'ghost'" [glyph]="'cart'"></button>
-<button fd-button label="Filter" [fdType]="'positive'" [glyph]="'filter'"></button>
-<button fd-button label="Filter" [fdType]="'negative'" [glyph]="'filter'"></button>
-<button fd-button label="Filter" [fdType]="'attention'" [glyph]="'filter'"></button>
-<button fd-button label="Filter" [fdType]="'transparent'" [glyph]="'filter'"></button>
+<button fd-button label="Add to Cart"  glyph="cart" ariaLabel="Add to Cart"></button>
```
Comment (droshev) [inline apps/docs/src/app/core/component-docs/button/examples/button-icons-example.component.html:1]: those updates of the doc where `[glyph]` is replaced with `glyph` should be done in separate PR

**Review by droshev — DISMISSED** [review]
i think the PR is mixing few things together

File: libs/core/src/lib/button/button.component.ts (hunk 3 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/button/button.component.ts:41]: @mikerodonnell89 
https://github.com/SAP/fundamental-ngx/pull/6385#pullrequestreview-742950708
requested for this change.

File: apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html (hunk 5 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html:2]: @droshev Previously a comment was shared saying if we are touching any file and that file has unwanted braces can be removed and this change was introduced as part of the previously shared comment.

**Comment by puru-hk** [comment]
All issues are fixed.

File: libs/core/src/lib/bar/button-bar/button-bar.component.ts (hunk 2 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/bar/button-bar/button-bar.component.ts:27]: this change is overridden when ariaLabel is passed or label is passed as a parameter.

File: libs/core/src/lib/bar/button-bar/button-bar.component.ts (hunk 1 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/bar/button-bar/button-bar.component.ts:26]: fixed

File: apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html (hunk 5 above)
Comment (droshev) [reply apps/docs/src/app/core/component-docs/button/examples/button-types-example.component.html:2]: I don't see why you would touch this file for the initial idea of your PR

**Comment by Lokanathan-k** [comment]
waiting for the feedback from the Ngx core and accessibility team.

**Comment by droshev** [comment]
@Lokanathan-k can we focus only on reading the button types(Emphasized and the semantic ones) the other types are just design types and they may confused the user (e.g. ghost button)?
So what we need is to have an internal logic to add the "Emphasized, warning, positive, and negative".

**Comment by Lokanathan-k** [comment]
> @Lokanathan-k can we focus only on reading the button types(Emphasized and the semantic ones) the other types are just design types and they may confused the user (e.g. ghost button)?
> So what we need is to have an internal logic to add the "Emphasized, warning, positive, and negative".

@droshev done, pls have a look

File: libs/core/src/lib/button/button.component.ts [hunk 8]
```diff
@@ -37,14 +37,21 @@ import { applyCssClass } from '@fundamental-ngx/core/utils';
...
+        '[attr.aria-label]': 'ariaLabel ? ariaLabel : specialButtonType.includes(fdType) ? label !==null ?  label +" ,"+ fdType : glyph !== undefined || glyph !== null ? fdType +","+ glyph.split("-").join(" ")  : title : title'
     }
 })
 export class ButtonComponent extends BaseButton implements OnChanges, CssClassBuilder, OnInit,  OnDestroy {
     /** The property allows user to pass additional css classes. */
     @Input()
     class = '';
 
+    /** @hidden */
+    specialButtonType: Array<string> = ['emphasized', 'positive', 'negative', 'attention'];
```
Comment (platon-rov) [inline libs/core/src/lib/button/button.component.ts:50]: Could you please care about supporting i18n.

File: libs/core/src/lib/button/button.component.ts [hunk 9]
```diff
@@ -37,14 +37,21 @@ import { applyCssClass } from '@fundamental-ngx/core/utils';
     changeDetection: ChangeDetectionStrategy.OnPush,
     host: {
         '[attr.type]': 'type',
-        '[attr.disabled]': '_disabled || null'
+        '[attr.disabled]': '_disabled || null',
+        '[attr.aria-label]': 'ariaLabel ? ariaLabel : specialButtonType.includes(fdType) ? label !==null ?  label +" ,"+ fdType : glyph !== undefined || glyph !== null ? fdType +","+ glyph.split("-").join(" ")  : title : title'
```
Comment (platon-rov) [inline libs/core/src/lib/button/button.component.ts:41]: Looks to hard to read.

Please use `!= null` for checking for not null and not undefined.

Instead of `a ? a : b` you can use `a || b`.

Please rewrite this expression.

**Review by platon-rov — DISMISSED** [review]

File: libs/core/src/lib/button/button.component.ts (hunk 8 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/button/button.component.ts:50]: this array is used to handle only button types that has to be read out by default., other button types are not required since it creates unwanted confusion this array.this is used to handle internally.

File: libs/core/src/lib/button/button.component.ts (hunk 9 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/button/button.component.ts:41]: @platon-rov this(a ? a : b you can use a || b) doesnt work as expected, when ariaLabel is passed the ariaLabel has to be retained. a || b doesn't do that.

File: libs/core/src/lib/button/button.component.ts (hunk 8 above)
Comment (platon-rov) [reply libs/core/src/lib/button/button.component.ts:50]: Got it, so maybe we need to rename the field to avoid such misunderstanding in feature. I guess we also has alll the button types as enum, do we? In that case we can use as items for that array enum values from there.

**Comment by puru-hk** [comment]
Issue 1: Not fixed
Issue 2: Fixed
Issue 3: Fixed
Issue 4: Fixed
Issue 5: Fixed

**Review by puru-hk — APPROVED** [review]
